src/hotspot/share/opto/gcm.cpp

```diff
@@ -770,7 +770,24 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
       // Add an anti-dep edge, and squeeze 'load' into the highest block.
       assert(store != load->find_exact_control(load->in(0)), "dependence cycle found");
       if (verify) {
-        assert(store->find_edge(load) != -1, "missing precedence edge");
+#ifdef ASSERT
+        // We expect an anti-dependence edge from 'load' to 'store', except when
+        // implicit_null_check() has hoisted 'store' above its early block to
+        // perform an implicit null check, and 'load' is placed in the null
+        // block. In this case it is safe to ignore the anti-dependence, as the
+        // null block is only reached if 'store' tries to write to null.
+        Node* store_null_check = store->find_out_with(Op_MachNullCheck);
+        Block* null_block = NULL;
+        if (store_null_check != NULL) {
+          Node* if_true = store_null_check->find_out_with(Op_IfTrue);
+          assert(if_true != NULL, "null check without null projection");
+          Node* null_block_region = if_true->find_out_with(Op_Region);
+          assert(null_block_region != NULL, "null check without null region");
+          null_block = get_block_for_node(null_block_region);
+        }
+#endif
+        assert((store_null_check != NULL && LCA == null_block) ||
```

Comment:
I think you could avoid the `store_null_check != NULL` check in the assert because `LCA` is always non-null.

Reply:
Done, thanks!